panic: runtime error: index out of range

Steps:

use http proxy such as cntlm on port 3328 

```
brew install annie 
annie -x http://localhost:3328 -i https://www.youtube.com/watch?v=ZtgzKBrU1GY

```

Result:
```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/iawia002/annie/extractors/youtube.decipherTokens(0xc00015e100, 0x5, 0x8, 0x0, 0x0, 0x0, 0x0)
	/private/tmp/annie-20190302-78152-y6yc8f/annie-0.9.3/src/github.com/iawia002/annie/extractors/youtube/signature.go:126 +0x3f7
github.com/iawia002/annie/extractors/youtube.getSig(0x0, 0x0, 0xc0001e80f0, 0x2d, 0x1659640, 0xc0004160de, 0x13805b1, 0x3)
	/private/tmp/annie-20190302-78152-y6yc8f/annie-0.9.3/src/github.com/iawia002/annie/extractors/youtube/youtube.go:50 +0x11e
github.com/iawia002/annie/extractors/youtube.genSignedURL(0xc000422400, 0x3a0, 0xc0001d73a8, 0xc0001e80f0, 0x2d, 0xc0001fa450, 0x3, 0x3, 0x1d)
	/private/tmp/annie-20190302-78152-y6yc8f/annie-0.9.3/src/github.com/iawia002/annie/extractors/youtube/youtube.go:66 +0x181
github.com/iawia002/annie/extractors/youtube.extractVideoURLS(0xc0001e8120, 0x27, 0xc000416000, 0x947c, 0xc000406000, 0xbfc, 0xc0001e80f0, 0x2d, 0x7ffeefbff7ca, 0x2b, ...)
	/private/tmp/annie-20190302-78152-y6yc8f/annie-0.9.3/src/github.com/iawia002/annie/extractors/youtube/youtube.go:192 +0x6be
github.com/iawia002/annie/extractors/youtube.youtubeDownload(0x7ffeefbff7ca, 0x2b, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/private/tmp/annie-20190302-78152-y6yc8f/annie-0.9.3/src/github.com/iawia002/annie/extractors/youtube/youtube.go:143 +0x33a
github.com/iawia002/annie/extractors/youtube.Extract(0x7ffeefbff7ca, 0x2b, 0x13835be, 0x7, 0x0, 0xc000120660, 0x3)
	/private/tmp/annie-20190302-78152-y6yc8f/annie-0.9.3/src/github.com/iawia002/annie/extractors/youtube/youtube.go:83 +0x69d
main.download(0x7ffeefbff7ca, 0x2b)
	/private/tmp/annie-20190302-78152-y6yc8f/annie-0.9.3/src/github.com/iawia002/annie/main.go:125 +0x699
main.main()
	/private/tmp/annie-20190302-78152-y6yc8f/annie-0.9.3/src/github.com/iawia002/annie/main.go:221 +0x156
```

panic: runtime error: index out of range                                                                                                                                                                    
                                                                                                                                                                                                            
goroutine 9 [running]:                                                                                                                                                                                      
github.com/iawia002/annie/extractors/youtube.decipherTokens(0xc0000797c0, 0x3, 0                                                                                                                            
x4, 0x0, 0x0, 0x0, 0x0)                                                                                                                                                                                     
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/                                                                                                                            
signature.go:126 +0x3db                                                                                                                                                                                     
github.com/iawia002/annie/extractors/youtube.getSig(0x0, 0x0, 0xc0001e4390, 0x2d                                                                                                                            
, 0x9c5620, 0x4114b1, 0xc0003ecb3a, 0x73ffaf)                                                                                                                                                               
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/                                                                                                                            
youtube.go:50 +0x13e                                                                                                                                                                                        
github.com/iawia002/annie/extractors/youtube.genSignedURL(0xc00019a380, 0x35b, 0                                                                                                                            
xc000252420, 0xc0001e4390, 0x2d, 0xc0002524e0, 0x3, 0x3, 0x440471)                                                                                                                                          
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/                                                                                                                            
youtube.go:66 +0x181                                                                                                                                                                                        
github.com/iawia002/annie/extractors/youtube.extractVideoURLS(0xc00028e400, 0x17                                                                                                                            
, 0xc0003eca80, 0x64c2, 0xc000194c00, 0xba0, 0xc0001e4390, 0x2d, 0xc000058180, 0                                                                                                                            
x53, ...)                                                                                                                                                                                                   
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/                                                                                                                            
youtube.go:192 +0x54a                                                                                                                                                                                       
github.com/iawia002/annie/extractors/youtube.youtubeDownload(0xc000058180, 0x53,                                                                                                                            
 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)                                                                                                                                                               
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/                                                                                                                            
youtube.go:143 +0x2fb                                                                                                                                                                                       
github.com/iawia002/annie/extractors/youtube.Extract.func1(0xc00005a2f0, 0x0, 0x                                                                                                                            
c000058180, 0x53, 0xc0000f4c80, 0x1c, 0x1c)                                                                                                                                                                 
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/                                                                                                                            
youtube.go:109 +0x7a                                                                                                                                                                                        
created by github.com/iawia002/annie/extractors/youtube.Extract                                                                                                                                             
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/                                                                                                                            
youtube.go:107 +0x63e

I am also getting this without proxies or anything.

```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/iawia002/annie/extractors/youtube.decipherTokens(0xc00028a180, 0x5, 0x8, 0x0, 0x0, 0x0, 0x0)
        /home/pentecostjonathan/go/src/github.com/iawia002/annie/extractors/youtube/signature.go:126 +0x3f7
github.com/iawia002/annie/extractors/youtube.getSig(0x0, 0x0, 0xc00028c030, 0x2d, 0xa5bc80, 0xc000446041, 0x78540d, 0x3)
        /home/pentecostjonathan/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:50 +0x11e
github.com/iawia002/annie/extractors/youtube.genSignedURL(0xc00044a400, 0x38b, 0xc000195398, 0xc00028c030, 0x2d, 0xc000278480, 0x3, 0x3, 0xd)
        /home/pentecostjonathan/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:66 +0x181
github.com/iawia002/annie/extractors/youtube.extractVideoURLS(0xc0002820a0, 0x4c, 0xc000446000, 0x3fe6, 0xc000430000, 0xc6a, 0xc00028c030, 0x2d, 0x7ffc8d1a1d4a, 0x2b, ...)
        /home/pentecostjonathan/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:192 +0x6c1
github.com/iawia002/annie/extractors/youtube.youtubeDownload(0x7ffc8d1a1d4a, 0x2b, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/pentecostjonathan/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:143 +0x33a
github.com/iawia002/annie/extractors/youtube.Extract(0x7ffc8d1a1d4a, 0x2b, 0x787ffe, 0x7, 0x1, 0xc0000873b0, 0x3)
        /home/pentecostjonathan/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:83 +0x69d
main.download(0x7ffc8d1a1d4a, 0x2b, 0x7a2c08)
        /home/pentecostjonathan/go/src/github.com/iawia002/annie/main.go:127 +0x6c0
main.main()
        /home/pentecostjonathan/go/src/github.com/iawia002/annie/main.go:233 +0x165
```

annie -s 127.0.0.1:1080 youtube视频地址
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/iawia002/annie/extractors/youtube.decipherTokens(0xc0001e8180, 0x5, 0x8, 0x0, 0x0, 0x0, 0x0)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/signature.go:126 +0x3e2
github.com/iawia002/annie/extractors/youtube.getSig(0x0, 0x0, 0xc0001ea030, 0x2d, 0x9b5d40, 0x411a48, 0xc0003840c5, 0x73c2c6)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:50 +0x145
github.com/iawia002/annie/extractors/youtube.genSignedURL(0xc000390380, 0x340, 0xc0001cc360, 0xc0001ea030, 0x2d, 0xc0001cc420, 0x3, 0x3, 0x43f828)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:66 +0x188
github.com/iawia002/annie/extractors/youtube.extractVideoURLS(0xc0001da240, 0x18, 0xc000384000, 0x5820, 0xc00038ac00, 0xb6e, 0xc0001ea030, 0x2d, 0xc000086030, 0x2b, ...)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:192 +0x551
github.com/iawia002/annie/extractors/youtube.youtubeDownload(0xc000086030, 0x2b, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:143 +0x302
github.com/iawia002/annie/extractors/youtube.Extract(0xc000086030, 0x2b, 0x73f269, 0x7, 0x0, 0x0, 0x0)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:83 +0x6ee
main.download(0xc000086030, 0x2b)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/main.go:125 +0x6e1
main.main()
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/main.go:221 +0x121

The tests are also failing.

```
$ go test -v -cover
=== RUN   TestYoutube
=== RUN   TestYoutube/normal_test
panic: runtime error: index out of range [recovered]
        panic: runtime error: index out of range

goroutine 7 [running]:
testing.tRunner.func1(0x1812320)
        /usr/local/go/src/testing/testing.go:830 +0x364
panic(0x2ac668, 0x51b280)
        /usr/local/go/src/runtime/panic.go:522 +0x1a0
github.com/iawia002/annie/test.Check(0x1812320, 0x2ef2ca, 0x2b, 0x2e7852, 0x17, 0x2eca37, 0x24, 0x0, 0x0, 0x2e60a5, ...)
        /home/zns/go/src/github.com/iawia002/annie/test/utils.go:33 +0x3f8
github.com/iawia002/annie/extractors/youtube.TestYoutube.func1(0x1812320)
        /home/zns/go/src/github.com/iawia002/annie/extractors/youtube/youtube_test.go:102 +0x1ac
testing.tRunner(0x1812320, 0x18107d8)
        /usr/local/go/src/testing/testing.go:865 +0xa4
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:916 +0x2b8
exit status 2
FAIL    github.com/iawia002/annie/extractors/youtube    18.635s
```

The algorithm may have changed as there are also errors reported in https://github.com/rylio/ytdl/issues/69. 

There may be a possible fix here: https://github.com/ytdl-org/youtube-dl/pull/18927

same here, 
annie -f 134 https://www.youtube.com/watch?v=ucZemSXcjtM
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/iawia002/annie/extractors/youtube.decipherTokens(0xc00010a100, 0x5, 0x8, 0x0, 0x0, 0x0, 0x0)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/signature.go:126 +0x3e2
github.com/iawia002/annie/extractors/youtube.getSig(0x0, 0x0, 0xc00000ccf0, 0x2d, 0x9b5d40, 0x411a48, 0xc00036e03d, 0x73c2c6)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:50 +0x145
github.com/iawia002/annie/extractors/youtube.genSignedURL(0xc00007aa80, 0x36a, 0xc0001a0480, 0xc00000ccf0, 0x2d, 0xc0001a0540, 0x3, 0x3, 0x43f828)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:66 +0x188
github.com/iawia002/annie/extractors/youtube.extractVideoURLS(0xc000014380, 0x64, 0xc00036e000, 0x4861, 0xc0000a9200, 0x436, 0xc00000ccf0, 0x2d, 0xc00000c2d0, 0x2b, ...)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:192 +0x551
github.com/iawia002/annie/extractors/youtube.youtubeDownload(0xc00000c2d0, 0x2b, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:143 +0x302
github.com/iawia002/annie/extractors/youtube.Extract(0xc00000c2d0, 0x2b, 0x73f269, 0x7, 0x0, 0x0, 0x0)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/extractors/youtube/youtube.go:83 +0x6ee
main.download(0xc00000c2d0, 0x2b)
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/main.go:125 +0x6e1
main.main()
        /Users/mac/iawia002/go/src/github.com/iawia002/annie/main.go:221 +0x121

--- a/extractors/miaopai/miaopai.go
+++ b/extractors/miaopai/miaopai.go
@@ -3,13 +3,16 @@ package miaopai
 import (
 	"encoding/json"
 	"fmt"
+	"math/rand"
+	"strings"
+	"time"
 
 	"github.com/iawia002/annie/downloader"
 	"github.com/iawia002/annie/request"
 	"github.com/iawia002/annie/utils"
 )
 
-type miaopai struct {
+type miaopaiData struct {
 	Data struct {
 		Description string `json:"description"`
 		MetaData    []struct {
@@ -20,18 +23,36 @@ type miaopai struct {
 	} `json:"data"`
 }
 
+func getRandomString(l int) string {
+	rand.Seed(time.Now().UnixNano())
+
+	s := make([]string, 0)
+	chars := []string{
+		"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "n", "m", "o", "p", "q", "r", "s", "t", "u", "v",
+		"w", "x", "y", "z", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
+	}
+	for i := 0; i < l; i++ {
+		s = append(s, chars[rand.Intn(len(chars)-1)])
+	}
+	return strings.Join(s, ",")
+}
+
 // Extract is the main function for extracting data
 func Extract(url string) ([]downloader.Data, error) {
 	id := utils.MatchOneOf(url, `/media/([^\./]+)`, `/show(?:/channel)?/([^\./]+)`)[1]
+
+	var data miaopaiData
 	jsonString, err := request.Get(
-		fmt.Sprintf("https://n.miaopai.com/api/aj_media/info.json?smid=%s", id), url, nil,
+		fmt.Sprintf("https://n.miaopai.com/api/aj_media/info.json?smid=%s&appid=530&_cb=_jsonp%s", id, getRandomString(10)),
+		url, nil,
 	)
 	if err != nil {
 		return downloader.EmptyList, err
 	}
-	var data miaopai
-	json.Unmarshal([]byte(jsonString), &data)
-
+	err = json.Unmarshal([]byte(jsonString), &data)
+	if err != nil {
+		return downloader.EmptyList, err
+	}
 	realURL := data.Data.MetaData[0].URLs.M
 	size, err := request.Size(realURL, url)
 	if err != nil {

--- a/extractors/miaopai/miaopai_test.go
+++ b/extractors/miaopai/miaopai_test.go
@@ -15,7 +15,7 @@ func TestDownload(t *testing.T) {
 		args test.Args
 	}{
 		{
-			name: "normal test 2",
+			name: "normal test",
 			args: test.Args{
 				URL:   "http://n.miaopai.com/media/Dqg5Pmb~I6lChdvOb-~r1BpKzzDu~MPr",
 				Title: "小学霸6点半起床学习:想赢在起跑线",
@@ -23,7 +23,7 @@ func TestDownload(t *testing.T) {
 			},
 		},
 		{
-			name: "normal test 3",
+			name: "normal test",
 			args: test.Args{
 				URL:   "https://n.miaopai.com/media/qVWj3dVK2oSxtW~vSq2tGeBKPE--tPSp",
 				Title: "如果你家的猫喜欢咬人怎么办？",

--- a/extractors/miaopai/miaopai_test.go
+++ b/extractors/miaopai/miaopai_test.go
@@ -9,23 +9,15 @@ import (
 
 func TestDownload(t *testing.T) {
 	config.InfoOnly = true
-	config.RetryTimes = 100
+	config.RetryTimes = 10
 	tests := []struct {
 		name string
 		args test.Args
 	}{
 		{
-			name: "normal test",
-			args: test.Args{
-				URL:   "https://www.miaopai.com/show/nPWJvdR4z2Bg1Sz3PJpNYffjpDgEiuv4msALgw__.htm",
-				Title: "情人节特辑：一个来自绝地求生的爱情故事，送给已经离开的你",
-				Size:  12135847,
-			},
-		},
-		{
 			name: "normal test 2",
 			args: test.Args{
-				URL:   "https://m.miaopai.com/show/channel/3PCuI5IZ6wdSZISmTtasYTa-l~wrVxk1yEgWRQ__",
+				URL:   "http://n.miaopai.com/media/Dqg5Pmb~I6lChdvOb-~r1BpKzzDu~MPr",
 				Title: "小学霸6点半起床学习:想赢在起跑线",
 				Size:  6743958,
 			},
@@ -33,7 +25,7 @@ func TestDownload(t *testing.T) {
 		{
 			name: "normal test 3",
 			args: test.Args{
-				URL:   "http://n.miaopai.com/media/qVWj3dVK2oSxtW~vSq2tGeBKPE--tPSp",
+				URL:   "https://n.miaopai.com/media/qVWj3dVK2oSxtW~vSq2tGeBKPE--tPSp",
 				Title: "如果你家的猫喜欢咬人怎么办？",
 				Size:  3615866,
 			},

--- a/extractors/netease/netease_test.go
+++ b/extractors/netease/netease_test.go
@@ -17,7 +17,7 @@ func TestDownload(t *testing.T) {
 			name: "mv test 1",
 			args: test.Args{
 				URL:   "https://music.163.com/#/mv?id=5547010",
-				Title: "There For You - Troye Sivan - 高清MV - 网易云音乐",
+				Title: "There For You",
 				Size:  24249078,
 			},
 		},
@@ -25,7 +25,7 @@ func TestDownload(t *testing.T) {
 			name: "video test 1",
 			args: test.Args{
 				URL:   "https://music.163.com/#/video?id=C8C9D11629798595BD28451DE3AC9FF4",
-				Title: "＃金曜日の新垣结衣 总集編〈全9編〉 - 视频 - 网易云音乐",
+				Title: "＃金曜日の新垣结衣 总集編〈全9編〉",
 				Size:  37408123,
 			},
 		},
@@ -33,7 +33,7 @@ func TestDownload(t *testing.T) {
 			name: "video test 2",
 			args: test.Args{
 				URL:   "https://music.163.com/m/video?id=6309CF62EF5D44FED5974536604944CF&userid=567080617",
-				Title: "当皮卡丘失去了小智就失去了全世界 - 视频 - 网易云音乐",
+				Title: "当皮卡丘失去了小智就失去了全世界",
 				Size:  28547736,
 			},
 		},

--- a/extractors/pornhub/pornhub.go
+++ b/extractors/pornhub/pornhub.go
@@ -33,10 +33,7 @@ func Extract(url string) ([]downloader.Data, error) {
 	realURLs := utils.MatchOneOf(html, `"mediaDefinitions":(.+?),"isVertical"`)
 
 	var pornhubs []pornhubData
-	err = json.Unmarshal([]byte(realURLs[1]), &pornhubs)
-	if err != nil {
-		return downloader.EmptyList, err
-	}
+	json.Unmarshal([]byte(realURLs[1]), &pornhubs)
 
 	streams := make(map[string]downloader.Stream, len(pornhubs))
 	for _, data := range pornhubs {

--- a/extractors/weibo/weibo_test.go
+++ b/extractors/weibo/weibo_test.go
@@ -9,19 +9,11 @@ import (
 
 func TestDownload(t *testing.T) {
 	config.InfoOnly = true
-	config.RetryTimes = 100
+	config.RetryTimes = 10
 	tests := []struct {
 		name string
 		args test.Args
 	}{
-		{
-			name: "normal test",
-			args: test.Args{
-				URL:   "https://m.weibo.cn/2815133121/G9VBqbsWM",
-				Title: "当你超过25岁再去夜店……",
-				Size:  3112080,
-			},
-		},
 		{
 			name: "fid url test",
 			args: test.Args{

--- a/extractors/yinyuetai/yinyuetai_test.go
+++ b/extractors/yinyuetai/yinyuetai_test.go
@@ -14,12 +14,12 @@ func TestDownload(t *testing.T) {
 		args test.Args
 	}{
 		{
-			name: "h5 test",
+			name: "normal test",
 			args: test.Args{
-				URL:     "http://v.yinyuetai.com/video/h5/820981",
-				Title:   "Rap God",
-				Size:    144401919,
-				Quality: "超清",
+				URL:     "http://v.yinyuetai.com/video/3386385",
+				Title:   "什么是爱/ What is Love",
+				Size:    20028736,
+				Quality: "流畅",
 			},
 		},
 	}

--- a/extractors/youtube/signature.go
+++ b/extractors/youtube/signature.go
@@ -1,89 +1,140 @@
 package youtube
 
 import (
-	"errors"
 	"fmt"
+	"net/url"
 	"regexp"
 	"strconv"
 	"strings"
+
+	"github.com/iawia002/annie/request"
 )
+
+func getDownloadURL(stream url.Values, htmlPlayerFile string) (string, error) {
+	var signature string
+	if s := stream.Get("s"); len(s) > 0 {
+		tokens, err := getSigTokens(htmlPlayerFile)
+		if err != nil {
+			return "", err
+		}
+		signature = decipherTokens(tokens, s)
+	} else {
+		if sig := stream.Get("sig"); len(sig) > 0 {
+			signature = sig
+		}
+	}
+	var urlString string
+	if s := stream.Get("url"); len(s) > 0 {
+		urlString = s
+	} else if s := stream.Get("stream"); len(s) > 0 {
+		if c := stream.Get("conn"); len(c) > 0 {
+			urlString = c
+			if urlString[len(urlString)-1] != '/' {
+				urlString += "/"
+			}
+		}
+		urlString += s
+	} else {
+		return "", fmt.Errorf("couldn't extract url from format")
+	}
+	urlString, err := url.QueryUnescape(urlString)
+	if err != nil {
+		return "", err
+	}
+	u, err := url.Parse(urlString)
+	if err != nil {
+		return "", err
+	}
+	query := u.Query()
+	query.Set("ratebypass", "yes")
+	if len(signature) > 0 {
+		if sp := stream.Get("sp"); sp != "" {
+			query.Set(sp, signature)
+		} else {
+			query.Set("signature", signature)
+		}
+	}
+	u.RawQuery = query.Encode()
+	return u.String(), nil
+}
 
 // The algorithm comes from https://github.com/rylio/ytdl, it's also MIT License
 // Many thanks
+
 const (
-	jsvarStr   = `[a-zA-Z_\$][a-zA-Z_0-9]*`
-	reverseStr = `:function\(a\)\{` +
-		`(?:return )?a\.reverse\(\)` +
-		`\}`
-	sliceStr = `:function\(a,b\)\{` +
-		`return a\.slice\(b\)` +
-		`\}`
-	spliceStr = `:function\(a,b\)\{` +
-		`a\.splice\(0,b\)` +
-		`\}`
-	swapStr = `:function\(a,b\)\{` +
-		`var c=a\[0\];a\[0\]=a\[b(?:%a\.length)?\];a\[b(?:%a\.length)?\]=c(?:;return a)?` +
-		`\}`
+	jsvarStr   = "[a-zA-Z_\\$][a-zA-Z_0-9]*"
+	reverseStr = ":function\\(a\\)\\{" +
+		"(?:return )?a\\.reverse\\(\\)" +
+		"\\}"
+	sliceStr = ":function\\(a,b\\)\\{" +
+		"return a\\.slice\\(b\\)" +
+		"\\}"
+	spliceStr = ":function\\(a,b\\)\\{" +
+		"a\\.splice\\(0,b\\)" +
+		"\\}"
+	swapStr = ":function\\(a,b\\)\\{" +
+		"var c=a\\[0\\];a\\[0\\]=a\\[b(?:%a\\.length)?\\];a\\[b(?:%a\\.length)?\\]=c(?:;return a)?" +
+		"\\}"
 )
 
 var actionsObjRegexp = regexp.MustCompile(fmt.Sprintf(
-	`var (%s)=\{((?:(?:%s%s|%s%s|%s%s|%s%s),?\n?)+)\};`,
+	"var (%s)=\\{((?:(?:%s%s|%s%s|%s%s|%s%s),?\\n?)+)\\};",
 	jsvarStr, jsvarStr, reverseStr, jsvarStr, sliceStr, jsvarStr, spliceStr, jsvarStr, swapStr,
 ))
 
 var actionsFuncRegexp = regexp.MustCompile(fmt.Sprintf(
-	`function(?: %s)?\(a\)\{`+
-		`a=a\.split\(""\);\s*`+
-		`((?:(?:a=)?%s\.%s\(a,\d+\);)+)`+
-		`return a\.join\(""\)`+
-		`\}`,
-	jsvarStr, jsvarStr, jsvarStr,
+	"function(?: %s)?\\(a\\)\\{"+
+		"a=a\\.split\\(\"\"\\);\\s*"+
+		"((?:(?:a=)?%s\\.%s\\(a,\\d+\\);)+)"+
+		"return a\\.join\\(\"\"\\)"+
+		"\\}", jsvarStr, jsvarStr, jsvarStr,
 ))
 
-var reverseRegexp = regexp.MustCompile(fmt.Sprintf(
-	`(?m)(?:^|,)(%s)%s`, jsvarStr, reverseStr,
-))
-var sliceRegexp = regexp.MustCompile(fmt.Sprintf(
-	`(?m)(?:^|,)(%s)%s`, jsvarStr, sliceStr,
-))
-var spliceRegexp = regexp.MustCompile(fmt.Sprintf(
-	`(?m)(?:^|,)(%s)%s`, jsvarStr, spliceStr,
-))
-var swapRegexp = regexp.MustCompile(fmt.Sprintf(
-	`(?m)(?:^|,)(%s)%s`, jsvarStr, swapStr,
-))
+var reverseRegexp = regexp.MustCompile(fmt.Sprintf("(?m)(?:^|,)(%s)%s", jsvarStr, reverseStr))
+var sliceRegexp = regexp.MustCompile(fmt.Sprintf("(?m)(?:^|,)(%s)%s", jsvarStr, sliceStr))
+var spliceRegexp = regexp.MustCompile(fmt.Sprintf("(?m)(?:^|,)(%s)%s", jsvarStr, spliceStr))
+var swapRegexp = regexp.MustCompile(fmt.Sprintf("(?m)(?:^|,)(%s)%s", jsvarStr, swapStr))
 
-func getSigTokens(html string) ([]string, error) {
-	objResult := actionsObjRegexp.FindStringSubmatch(html)
-	funcResult := actionsFuncRegexp.FindStringSubmatch(html)
+func getSigTokens(htmlPlayerFile string) ([]string, error) {
+	u, _ := url.Parse("https://www.youtube.com/watch")
+	p, err := url.Parse(htmlPlayerFile)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := request.Get(u.ResolveReference(p).String(), referer, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	objResult := actionsObjRegexp.FindStringSubmatch(body)
+	funcResult := actionsFuncRegexp.FindStringSubmatch(body)
 
 	if len(objResult) < 3 || len(funcResult) < 2 {
-		return nil, errors.New("error parsing signature tokens")
+		return nil, fmt.Errorf("error parsing signature tokens")
 	}
-	obj := strings.Replace(objResult[1], "$", `\$`, -1)
-	objBody := strings.Replace(objResult[2], "$", `\$`, -1)
-	funcBody := strings.Replace(funcResult[1], "$", `\$`, -1)
+	obj := strings.Replace(objResult[1], "$", "\\$", -1)
+	objBody := strings.Replace(objResult[2], "$", "\\$", -1)
+	funcBody := strings.Replace(funcResult[1], "$", "\\$", -1)
 
 	var reverseKey, sliceKey, spliceKey, swapKey string
 	var result []string
 
 	if result = reverseRegexp.FindStringSubmatch(objBody); len(result) > 1 {
-		reverseKey = strings.Replace(result[1], "$", `\$`, -1)
+		reverseKey = strings.Replace(result[1], "$", "\\$", -1)
 	}
 	if result = sliceRegexp.FindStringSubmatch(objBody); len(result) > 1 {
-		sliceKey = strings.Replace(result[1], "$", `\$`, -1)
+		sliceKey = strings.Replace(result[1], "$", "\\$", -1)
 	}
 	if result = spliceRegexp.FindStringSubmatch(objBody); len(result) > 1 {
-		spliceKey = strings.Replace(result[1], "$", `\$`, -1)
+		spliceKey = strings.Replace(result[1], "$", "\\$", -1)
 	}
 	if result = swapRegexp.FindStringSubmatch(objBody); len(result) > 1 {
-		swapKey = strings.Replace(result[1], "$", `\$`, -1)
+		swapKey = strings.Replace(result[1], "$", "\\$", -1)
 	}
 
 	keys := []string{reverseKey, sliceKey, spliceKey, swapKey}
-	regex, err := regexp.Compile(fmt.Sprintf(
-		`(?:a=)?%s\.(%s)\(a,(\d+)\)`, obj, strings.Join(keys, "|"),
-	))
+	regex, err := regexp.Compile(fmt.Sprintf("(?:a=)?%s\\.(%s)\\(a,(\\d+)\\)", obj, strings.Join(keys, "|")))
 	if err != nil {
 		return nil, err
 	}

--- a/extractors/youtube/youtube.go
+++ b/extractors/youtube/youtube.go
@@ -31,51 +31,6 @@ type youtubeData struct {
 
 const referer = "https://www.youtube.com"
 
-var tokensCache = make(map[string][]string)
-
-func getSig(sig, js string) (string, error) {
-	sigURL := fmt.Sprintf("https://www.youtube.com%s", js)
-	tokens, ok := tokensCache[sigURL]
-	if !ok {
-		html, err := request.Get(sigURL, referer, nil)
-		if err != nil {
-			return "", err
-		}
-		tokens, err = getSigTokens(html)
-		if err != nil {
-			return "", err
-		}
-		tokensCache[sigURL] = tokens
-	}
-	return decipherTokens(tokens, sig), nil
-}
-
-func genSignedURL(streamURL string, stream url.Values, js string) (string, error) {
-	var (
-		realURL, sig string
-		err          error
-	)
-	if strings.Contains(streamURL, "signature=") {
-		// URL itself already has a signature parameter
-		realURL = streamURL
-	} else {
-		// URL has no signature parameter
-		sig = stream.Get("sig")
-		if sig == "" {
-			// Signature need decrypt
-			sig, err = getSig(stream.Get("s"), js)
-			if err != nil {
-				return "", err
-			}
-		}
-		realURL = fmt.Sprintf("%s&signature=%s", streamURL, sig)
-	}
-	if !strings.Contains(realURL, "ratebypass") {
-		realURL += "&ratebypass=yes"
-	}
-	return realURL, nil
-}
-
 // Extract is the main function for extracting data
 func Extract(uri string) ([]downloader.Data, error) {
 	var err error
@@ -128,7 +83,7 @@ func youtubeDownload(uri string) downloader.Data {
 		return downloader.EmptyData(uri, errors.New("can't find vid"))
 	}
 	videoURL := fmt.Sprintf(
-		"https://www.youtube.com/watch?v=%s&gl=US&hl=en&has_verified=1&bpctr=9999999999",
+		"https://www.youtube.com/watch?v=%s",
 		vid[1],
 	)
 	html, err := request.Get(videoURL, referer, nil)
@@ -189,7 +144,7 @@ func extractVideoURLS(data youtubeData, referer string) (map[string]downloader.S
 		} else {
 			ext = utils.MatchOneOf(streamType, `(\w+)/(\w+);`)[2]
 		}
-		realURL, err := genSignedURL(stream.Get("url"), stream, data.Assets.JS)
+		realURL, err := getDownloadURL(stream, data.Assets.JS)
 		if err != nil {
 			return nil, err
 		}

--- a/extractors/youtube/youtube_test.go
+++ b/extractors/youtube/youtube_test.go
@@ -54,10 +54,10 @@ func TestYoutube(t *testing.T) {
 		{
 			name: "playlist test",
 			args: test.Args{
-				URL:     "https://www.youtube.com/watch?v=x2nKigmfzLQ&list=PLfyyPldyYFapPfxZOay2AoCCaNE1Ezd4_",
-				Title:   "【仔細看看】《羅根》黑白版-漫畫電影新高峰?! - 超粒方",
-				Size:    192310804,
-				Quality: `1080p video/webm; codecs="vp9"`,
+				URL:     "https://www.youtube.com/watch?v=Lt2pwLxJxgA&list=PLIYAO-qLriEtYm7UcXPH3SOJxgqjwRrIw",
+				Title:   "papi酱 - 你有酱婶儿的朋友吗？",
+				Size:    13549971,
+				Quality: `720p video/mp4; codecs="avc1.4d401f"`,
 			},
 			playlist: true,
 		},

--- a/extractors/youtube/youtube_test.go
+++ b/extractors/youtube/youtube_test.go
@@ -35,6 +35,15 @@ func TestYoutube(t *testing.T) {
 			},
 		},
 		{
+			name: "signature test",
+			args: test.Args{
+				URL:     "https://www.youtube.com/watch?v=ZtgzKBrU1GY",
+				Title:   "Halo Infinite - E3 2019 - Discover Hope",
+				Size:    877713473,
+				Quality: `2160p60 video/webm; codecs="vp9"`,
+			},
+		},
+		{
 			name: "normal test",
 			args: test.Args{
 				URL:     "https://www.youtube.com/watch?v=ASPku-eAZYs",

--- a/test/utils.go
+++ b/test/utils.go
@@ -25,9 +25,9 @@ func CheckData(args, data Args) bool {
 // Check check the result
 func Check(t *testing.T, args Args, data downloader.Data) {
 	// get the default stream
-	sortedStreams := make([]downloader.Stream, len(data.Streams))
-	for _, data := range data.Streams {
-		sortedStreams = append(sortedStreams, data)
+	sortedStreams := make([]downloader.Stream, 0, len(data.Streams))
+	for _, s := range data.Streams {
+		sortedStreams = append(sortedStreams, s)
 	}
 	sort.Slice(sortedStreams, func(i, j int) bool { return sortedStreams[i].Size > sortedStreams[j].Size })
 	defaultData := sortedStreams[0]

